galley/pkg/config/source/priority/source.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package priority
+
+import (
+	"sync"
+
+	"istio.io/istio/galley/pkg/config/collection"
+	"istio.io/istio/galley/pkg/config/event"
+)
+
+// Source is a processor.Source implementation that combines multiple sources in priority order
+// Such that sources later in the input list take priority over sources earlier in the list
+type Source struct {
+	mu      sync.Mutex
+	started bool
+
+	inputs  []event.Source
+	handler event.Handler
+
+	eventStateMu    sync.Mutex
+	eventPriorities map[string]int
+	fullSyncCounts  map[collection.Name]int
+}
+
+type priorityHandler struct {
+	priority int
+	src      *Source
+}
+
+var _ event.Source = &Source{}
+
+var (
+	discardHandler = event.SentinelHandler()
+)
+
+// New returns a new priority source, based on given input sources.
+func New(sources ...event.Source) *Source {
+	return &Source{
+		inputs:          sources,
+		eventPriorities: make(map[string]int),
+		fullSyncCounts:  make(map[collection.Name]int),
+	}
+}
+
+// Handle implements event.Handler
+func (ph *priorityHandler) Handle(e event.Event) {
+	ph.src.eventStateMu.Lock()
+	defer ph.src.eventStateMu.Unlock()
+
+	if e.Kind == event.FullSync {
+		ph.handleFullSync(e)
+	} else {
+		ph.handleEvent(e)
+	}
+}
+
+// handleFullSync handles FullSync events, which are a special case.
+// For each collection, we want to only send this once, after all upstream sources have sent theirs.
+func (ph *priorityHandler) handleFullSync(e event.Event) {
+	ph.src.fullSyncCounts[e.Source]++
+	if ph.src.fullSyncCounts[e.Source] == len(ph.src.inputs) {
+		ph.src.handler.Handle(e)
+	} else {
+		discardHandler.Handle(e)
```

Comment:
This seems redundant?

Reply:
Yeah. At one point I wondered if unhandled events were something I needed to avoid.